staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go

```diff
@@ -0,0 +1,544 @@
+					case "update":
+						_, err = v1client.Update(obj, metav1.UpdateOptions{}, subresource)
+					case "patch":
+						_, err = v1client.Patch(obj.GetName(), types.MergePatchType, []byte(`{"metadata":{"annotations":{"patch":"true"}}}`), metav1.PatchOptions{}, subresource)
+					}
+
+					if err == nil {
+						t.Errorf("expected error with message %s, but got no error", message)
+					} else if !strings.Contains(err.Error(), message) {
+						t.Errorf("expected error with message %s, but got %v", message, err)
+					}
+				})
+			}
+		}
+	}
+}
+
+func noopConverter(desiredAPIVersion string, obj runtime.RawExtension) (runtime.RawExtension, error) {
+	u := &unstructured.Unstructured{Object: map[string]interface{}{}}
+	if err := json.Unmarshal(obj.Raw, u); err != nil {
+		return runtime.RawExtension{}, fmt.Errorf("Fail to deserialize object: %s with error: %v", string(obj.Raw), err)
+	}
+	u.Object["apiVersion"] = desiredAPIVersion
+	raw, err := json.Marshal(u)
+	if err != nil {
+		return runtime.RawExtension{}, fmt.Errorf("Fail to serialize object: %v with error: %v", u, err)
+	}
+	return runtime.RawExtension{Raw: raw}, nil
+}
+
+func emptyResponseConverter(review apiextensionsv1beta1.ConversionReview) (apiextensionsv1beta1.ConversionReview, error) {
+	review.Response = &apiextensionsv1beta1.ConversionResponse{
+		UID:              review.Request.UID,
+		ConvertedObjects: []runtime.RawExtension{},
+		Result:           metav1.Status{Status: "Success"},
+	}
+	return review, nil
+}
+
+func failureResponseConverter(message string) func(review apiextensionsv1beta1.ConversionReview) (apiextensionsv1beta1.ConversionReview, error) {
+	return func(review apiextensionsv1beta1.ConversionReview) (apiextensionsv1beta1.ConversionReview, error) {
+		review.Response = &apiextensionsv1beta1.ConversionResponse{
+			UID:              review.Request.UID,
+			ConvertedObjects: []runtime.RawExtension{},
+			Result:           metav1.Status{Message: message, Status: "Failure"},
+		}
+		return review, nil
+	}
+}
+
+func nontrivialConverter(desiredAPIVersion string, obj runtime.RawExtension) (runtime.RawExtension, error) {
+	u := &unstructured.Unstructured{Object: map[string]interface{}{}}
+	if err := json.Unmarshal(obj.Raw, u); err != nil {
+		return runtime.RawExtension{}, fmt.Errorf("Fail to deserialize object: %s with error: %v", string(obj.Raw), err)
+	}
+
+	currentAPIVersion := u.Object["apiVersion"]
+	if currentAPIVersion == "v1beta2" && desiredAPIVersion == "v1beta1" {
+		u.Object["num"] = u.Object["numv2"]
+		u.Object["content"] = u.Object["contentv2"]
+		delete(u.Object, "v2only")
+		delete(u.Object, "numv2")
+		delete(u.Object, "contentv2")
+	}
+	if currentAPIVersion == "v1beta1" && desiredAPIVersion == "v1beta2" {
+		u.Object["numv2"] = u.Object["num"]
+		u.Object["contentv2"] = u.Object["content"]
+		u.Object["v2only"] = map[string]string{"newv2stuff": "hasadefault"}
```

Comment:
nit: this doesn't roundtrip

Reply:
Thanks for the reminder, I was going to experiment with what happens if a converter is not symmetrical, but never got around to it. I'll remove this.